src/twiliosim.cr

```diff
@@ -31,85 +36,155 @@ class Twiliosim::Server
-      from = params["From"]
-      unless params.has_key?("From")
-        plain_response(context, 400, "Body param 'To' is missing")
+      if response_body.blank?
         return
       end
-      to = params["To"]
-      context.response.status_code = 201
-      context.response.content_type = "application/json"
-      response = {"sid" => UUID.random().to_s()}
-      response.to_json(context.response)
-      spawn do
-        sleep 1.seconds
-        request_params = {"AccountSid" => account_sid, "From" => from, "To" => to, "CallStatus" => "in-progress"}
-        request_body = HTTP::Params.encode(request_params)
-        HTTP::Client.post(verboice_url, body: request_body) do |response|
-          response_body = response.body_io.gets
-          unless response_body
-            puts "Callback failed (body response is empty) - POST #{verboice_url} #{request_body} - #{response.status_code} - #{response.status_message}"
-            next
-          end
-          if response_body.blank?
-            puts "Callback failed (body response is empty) - POST #{verboice_url} #{request_body} - #{response.status_code} - #{response.status_message}"
-            next
-          end
-          case response_body
-          when %r(<Redirect>(.*)<\/Redirect>)
-            redirect_url = $1
-            unless redirect_url
-              puts "Callback failed (redirect url is missing) - POST #{verboice_url} #{request_body} - #{response.status_code} - #{response.status_message}"
-              next
-            end
-            if %r(<Say language="en">hangup<\/Say>).matches?(response_body)
-              spawn do
-                sleep 5.seconds
-                request_params["CallStatus"] = "completed"
-                request_body = HTTP::Params.encode(request_params)
-                HTTP::Client.post(redirect_url, body: request_body)
-              end
-            end
-          else
-            puts "Callback failed (redirect url is missing) - POST #{verboice_url} #{request_body} - #{response.status_code} - #{response.status_message}"
-            next
-          end
-        end
-      end
-    else
-      plain_response(context, 404, "404 Not Found")
+      response_body
     end
   end
-end
 
-private def plain_response(context, code, text)
-  context.response.status_code = code
-  context.response.content_type = "text/plain"
-  context.response.puts text
+  private def parse_ao_message(response_body : String) : TwilioAOMessage | Nil
+    if %r(<Say .+>(.+)<\/Say>.*<Redirect>(.+)<\/Redirect>).match(response_body)
+      message = $1
+      redirect_url = $2
+      TwilioAOMessage.new(message, redirect_url)
+    end
+  end
+
+  private def ao_message_redirect_url(ao_message : TwilioAOMessage)
+    ao_message.redirect_url
+  end
+
+  private def perform_response(reply_command : HangUp, call : TwilioCall) : ReplyCommand | Nil
+    redirect_url = ao_message_redirect_url(reply_command.ao_message)
+    return unless redirect_url
+    call.finish()
```

Comment:
Is this right? sholdn't the call be finished if there is no `redirect_url`?

Reply:
You're right. I hadn't' considered this alternative flow. I thought the compiler would need checking if the `redirect_url` exists, but it doesn't, so I'm removing it.

I'm pretty sure this flow isn't even possible because we're ignoring Verboice instructions without `redirect_url` [here](https://github.com/instedd/ncd_twilio_simulator/blob/7d24fae15fcbd5ecef15e3e835517fcc9dec36ff/src/twiliosim.cr#L140-L146)